src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/W3CPropagator.cs

```diff
@@ -0,0 +1,741 @@
+
+            do
+            {
+                int entrySeparator = baggageSpan.IndexOf(Comma);
+                ReadOnlySpan<char> currentEntry = entrySeparator >= 0 ? baggageSpan.Slice(0, entrySeparator) : baggageSpan;
+
+                int keyValueSeparator = currentEntry.IndexOf(Equal);
+                if (keyValueSeparator <= 0 || keyValueSeparator >= currentEntry.Length - 1)
+                {
+                    break; // invalid format
+                }
+
+                ReadOnlySpan<char> keySpan = currentEntry.Slice(0, keyValueSeparator);
+                ReadOnlySpan<char> valueSpan = currentEntry.Slice(keyValueSeparator + 1);
+
+                if (TryDecodeBaggageKey(keySpan, out string? key) && TryDecodeBaggageValue(valueSpan, out string value))
+                {
+                    baggageList ??= new List<KeyValuePair<string, string?>>();
+                    baggageList.Add(new KeyValuePair<string, string?>(key, value));
+                }
+
+                baggageSpan = entrySeparator >= 0 ? baggageSpan.Slice(entrySeparator + 1) : ReadOnlySpan<char>.Empty;
+            } while (baggageSpan.Length > 0);
+
+            baggage = baggageList;
+            return baggageList != null;
+        }
+
+        // list  = list-member 0*31( OWS "," OWS list-member )
+        // list-member = (key "=" value) / OWS
+        //
+        // key = ( lcalpha / DIGIT ) 0*255 ( keychar )
+        // keychar    = lcalpha / DIGIT / "_" / "-"/ "*" / "/" / "@"
+        // lcalpha    = %x61-7A ; a-z
+        //
+        // value    = 0*255(chr) nblk-chr
+        // nblk-chr = %x21-2B / %x2D-3C / %x3E-7E
+        // chr      = %x20 / nblk-chr
+
+        internal static string? ValidateTraceState(string? traceState)
+        {
+            if (string.IsNullOrEmpty(traceState))
+            {
+                return null; // invalid format
+            }
+
+            int processed = 0;
+
+            while (processed < traceState.Length)
+            {
+                ReadOnlySpan<char> traceStateSpan = traceState.AsSpan(processed);
+                int commaIndex = traceStateSpan.IndexOf(Comma);
+                ReadOnlySpan<char> entry = commaIndex >= 0 ? traceStateSpan.Slice(0, commaIndex) : traceStateSpan;
+                int delta = entry.Length + (commaIndex >= 0 ? 1 : 0); // +1 for the comma
+
+                if (processed + delta > MaxTraceStateEncodedLength)
+                {
+                    break; // entry exceeds max length
+                }
+
+                int equalIndex = entry.IndexOf(Equal);
+                if (equalIndex <= 0 || equalIndex >= entry.Length - 1)
+                {
+                    break; // invalid format
+                }
+
+                if (IsInvalidTraceStateKey(Trim(entry.Slice(0, equalIndex))) || IsInvalidTraceStateValue(TrimSpaceOnly(entry.Slice(equalIndex + 1))))
+                {
+                    break; // entry exceeds max length or invalid key/value, skip current entry
```

Comment:
The comment says skip current entry but the code appears to skip all entries after this one right? Might want to adjust the comment to match the behavior.

Reply:
Fixed.